src/Task.cc

```diff
@@ -1295,21 +1326,34 @@ void Task::resume_execution(ResumeRequest how, WaitRequest wait_how,
              << ptrace_req_name<NativeArch>(how)
              << (sig ? string(", signal ") + signal_name(sig) : string())
              << " tick_period " << tick_period << " wait " << wait_how;
-  address_of_last_execution_resume = ip();
-  how_last_execution_resumed = how;
   set_x86_debug_status(0);
 
   if (is_singlestep_resume(how)) {
     work_around_KNL_string_singlestep_bug();
-    singlestepping_instruction = trapped_instruction_at(this, ip());
-    if (singlestepping_instruction == TrappedInstruction::CPUID) {
-      // In KVM virtual machines (and maybe others), singlestepping over CPUID
-      // executes the following instruction as well. Work around that.
-      did_set_breakpoint_after_cpuid =
-        vm()->add_breakpoint(ip() + trapped_instruction_len(singlestepping_instruction), BKPT_INTERNAL);
+    if (is_x86ish(arch())) {
+      singlestepping_instruction = trapped_instruction_at(this, ip());
+      if (singlestepping_instruction == TrappedInstruction::CPUID) {
+        // In KVM virtual machines (and maybe others), singlestepping over CPUID
+        // executes the following instruction as well. Work around that.
+        did_set_breakpoint_after_cpuid =
+          vm()->add_breakpoint(ip() + trapped_instruction_len(singlestepping_instruction), BKPT_INTERNAL);
+      }
+    } else if (arch() == aarch64 && is_singlestep_resume(how_last_execution_resumed)) {
+      // On aarch64, if the last execution was any sort of single step, then
+      // resuming again with PTRACE_(SYSEMU_)SINGLESTEP will cause a debug fault
+      // upon exiting from the current instruction. To prevent this, we must
```

Comment:
I think "cause a debug fault before executing the next instruction" is clearer. "exiting an instruction" is not really a well-defined concept, at least in my head.

Reply:
Fair point.